SampleApplications/SDK/Opc.Ua.Server/NodeManager/MasterNodeManager.cs

```diff
@@ -2647,37 +2695,359 @@ protected static ServiceResult ValidateMonitoredItemModifyRequest(MonitoredItemM
+        /// <returns></returns>
+        protected ServiceResult ValidateWriteRequest(OperationContext operationContext, WriteValue writeValue)
+        {
+            ServiceResult serviceResult = WriteValue.Validate(writeValue);
+
+            if (ServiceResult.IsGood(serviceResult))
+            {
+                PermissionType requestedPermission = PermissionType.WriteAttribute;  //any attribute other than Value, RolePermissions or Historizing
+                if (writeValue.AttributeId == Attributes.RolePermissions)
+                {
+                    requestedPermission = PermissionType.WriteRolePermissions;
+                }
+                else if (writeValue.AttributeId == Attributes.Historizing)
+                {
+                    requestedPermission = PermissionType.WriteHistorizing;
+                }
+                else if (writeValue.AttributeId == Attributes.Value)
+                {
+                    requestedPermission = PermissionType.Write;
+                }
+
+                // check access rights and permissions
+                serviceResult = ValidatePermissions(operationContext, writeValue.NodeId, requestedPermission);
+            }
+            return serviceResult;
+        }
+
+        /// <summary>
+        /// Validates a HistoryRead request. It validates also access rights and role permissions
+        /// </summary>
+        /// <param name="operationContext"></param>
+        /// <param name="historyReadValueId"></param>
+        /// <returns></returns>
+        protected ServiceResult ValidateHistoryReadRequest(OperationContext operationContext, HistoryReadValueId historyReadValueId)
+        {
+            ServiceResult serviceResult = HistoryReadValueId.Validate(historyReadValueId);
+
+            if (ServiceResult.IsGood(serviceResult))
+            {
+                // check access rights and permissions
+                serviceResult = ValidatePermissions(operationContext, historyReadValueId.NodeId, PermissionType.ReadHistory);
+            }
+            return serviceResult;
         }
+
+        /// <summary>
+        ///  Validates a HistoryUpdate request. It validates also access rights and role permissions
+        /// </summary>
+        /// <param name="operationContext"></param>
+        /// <param name="historyUpdateDetails"></param>
+        /// <returns></returns>
+        protected ServiceResult ValidateHistoryUpdateRequest(OperationContext operationContext, HistoryUpdateDetails historyUpdateDetails)
+        {
+            ServiceResult serviceResult = HistoryUpdateDetails.Validate(historyUpdateDetails);
+
+            if (ServiceResult.IsGood(serviceResult))
+            {
+                // check access rights and permissions
+                serviceResult = ValidatePermissions(operationContext, historyUpdateDetails.NodeId, PermissionType.ModifyHistory);
+            }
+
+            return serviceResult;
+        }
+
+        #region Validate Permissions Methods
```

Comment:
>#region  [](start = 8, length = 8)

no #endregion?

Reply:
"Validate Permissions Methods" is an inner region of "#region Protected Members"